### PR TITLE
Include missing test files in runtests.jl

### PR DESCRIPTION
**Pull request by jlumpe**
There were several files in the `test/` directory that were not included in test sets in `test/runtests.jl`. I'm mostly interested in restoring `diffeqfunction_tests.jl` so I can add to it but figured I should do the rest at the same time.

`internal_euler_test.jl` and `traits_tests.jl` both cause errors. Maybe they are obsolete and should be removed entirely but I just commented them out in `runtests.jl` to make it obvious they were not being run.

I also moved "destats_tests.jl" to the `downstream/` directory because it imports `OrdinaryDiffEq`.

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
The traits tests should just be removed. This:

https://github.com/JuliaDiffEq/DiffEqBase.jl/blob/master/test/internal_euler_test.jl#L4-L6

should be updated to the "newer" syntax and use ODEFunction.

**Comment by jlumpe** [comment]
Fixed internal Euler test, also required some fixes in implementation as well. Refactored DiffEqFunction test file and added tests for `SplitFunction` and `DynamicalODEFunction`. This should now catch any problems created by implementing #233 or #284 instead of breaking things downstream.

File: test/internal_euler_test.jl [hunk 1]
```diff
@@ -1,9 +1,11 @@
 using DiffEqBase, DiffEqBase.InternalEuler
 
 # Try it
-ff(t,u) = u
-ff(::Val{:jac}, t, u) = 1
-ff(::Type{Val{:analytic}},t,u0) = u0*exp(t)
+ff = ODEFunction(
+    (u, p, t) -> u,
+    jac=(u, p, t) -> u,
```
Comment (devmotion) [inline test/internal_euler_test.jl]: The jacobian should be `(u, p, t) -> 1`.

Comment (ChrisRackauckas) [inline test/internal_euler_test.jl]:
```suggestion
    jac=(u, p, t) -> 1,
```

Comment (jlumpe) [reply test/internal_euler_test.jl]: Fixed.

Comment (jlumpe) [reply test/internal_euler_test.jl]: Fixed.

**Pull request merged**